### PR TITLE
daml-lf, kvutils: Move Daml-LF transaction and value handling logic to `kv-support` [KVL-1168]

### PR DESCRIPTION
**Pull request by hubert-da**
* Moves two functions from `KeyValueCommitting` and `TransactionCommitter` to `kv-support` which allows to stop depending on `//daml-lf/transaction:transaction_proto_java` and `//daml-lf/transaction:value_proto_java` in `kvutils`.
* Adds error handling.
* Adds new tests.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala [hunk 1]
```diff
@@ -92,9 +95,157 @@ object TransactionConversions {
       .sequence_
       .map(_ => RawTransaction(transactionBuilder.build.toByteString))
   }
+
+  def decodeContractIdsAndKeys(
```
Comment (cocreature) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:99]: The name here isn’t all that helpful. I’d suggest renaming it to something that contains `outputs` and adding a doc comment.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala [hunk 2]
```diff
@@ -92,9 +95,157 @@ object TransactionConversions {
...
+                        for {
+                          newContractIdsOrKeys <- TransactionCoder
+                            .nodeKey(nodeVersion, protoExercise)
+                            .map {
+                              case Some(key) => contractIdsOrKeys + ContractIdOrKey.Key(key)
+                              case None => contractIdsOrKeys
+                            }
+                          contractId <- ValueCoder.CidDecoder
+                            .decode(protoExercise.getContractIdStruct)
+                        } yield newContractIdsOrKeys + ContractIdOrKey.Id(contractId)
```
Comment (cocreature) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:148]: I don’t understand the intended semantics here. Looking at the usesite I thought this was about outputs. But here you are including contract ids of non-consuming exercises which doesn’t seem like something that should be an output.

Comment (cocreature) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:148]: Oh looking below I guess it’s because it can cause divulgence like the fetch.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala [hunk 3]
```diff
@@ -92,9 +95,157 @@ object TransactionConversions {
...
+                            .decode(protoCreate.getContractIdStruct)
+                        } yield newContractIdsOrKeys + ContractIdOrKey.Id(contractId)
+
+                      case NodeTypeCase.EXERCISE =>
+                        val protoExercise = node.getExercise
+                        for {
+                          newContractIdsOrKeys <- TransactionCoder
+                            .nodeKey(nodeVersion, protoExercise)
+                            .map {
+                              case Some(key) => contractIdsOrKeys + ContractIdOrKey.Key(key)
```
Comment (cocreature) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:143]: Similar question as for contract ids: Why as a key an output for a non-consuming exercise? Divulgence doesn’t seem to apply here.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala [hunk 4]
```diff
@@ -92,9 +95,157 @@ object TransactionConversions {
...
+                    }
+                  }
+                case (Left(error), _) => Left(error)
+              }
+          }
+          .left
+          .map(ConversionError.DecodeError)
+    }
+
+  def filterCreateAndExerciseNodes(
```
Comment (cocreature) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:172]: The name here doesn’t tell you if we’re keeping the creates & exercise (and only those) or if those are the ones that are filtered out. I’d suggest renaming it (e.g. `keepCreateAndExerciseNodes`, `) and adding a doc comment.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala [hunk 5]
```diff
@@ -92,9 +95,157 @@ object TransactionConversions {
...
+    Try(TransactionOuterClass.Transaction.parseFrom(rawTransaction.byteString)) match {
+      case Failure(throwable) => Left(ConversionError.ParseError(throwable.getMessage))
+      case Success(transaction) =>
+        val nodes = transaction.getNodesList.asScala
+        val nodeMap: Map[String, TransactionOuterClass.Node] =
+          nodes.view.map(n => n.getNodeId -> n).toMap
+
+        @tailrec
+        def goNodesToKeep(
+            todo: List[String],
```
Comment (cocreature) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:184]: Maybe worth considering using a FrontStack here.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala [hunk 6]
```diff
@@ -92,9 +95,157 @@ object TransactionConversions {
...
+                    goNodesToKeep(tail, result)
+                }
+              case None => Left(ConversionError.InternalError(s"Invalid transaction node id $head"))
+            }
+        }
+
+        goNodesToKeep(transaction.getRootsList.asScala.toList, Set.empty).map { nodesToKeep =>
+          val filteredRoots = transaction.getRootsList.asScala.filter(nodesToKeep)
+
+          def stripUnnecessaryNodes(node: TransactionOuterClass.Node): TransactionOuterClass.Node =
```
Comment (cocreature) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:210]: maybe move that somewhere out. Having a definition nested this deeply is a bit hard to follow.

File: daml-lf/kv-support/src/test/scala/com/daml/lf/kv/transactions/TransactionConversionsSpec.scala [hunk 7]
```diff
@@ -89,6 +94,268 @@ class TransactionConversionsSpec extends AnyWordSpec with Matchers {
...
+      builder.add(exerciseNode)
+      val result = TransactionConversions.decodeContractIdsAndKeys(
+        TransactionConversions
+          .encodeTransaction(builder.build())
+          .fold(error => fail(error.errorMessage), identity)
+      )
+      result shouldBe Right(Set(ContractIdOrKey.Id(exerciseNode.targetCoid)))
+    }
+
+    "return two outputs for an exercise with a key" in {
```
Comment (cocreature) [inline daml-lf/kv-support/src/test/scala/com/daml/lf/kv/transactions/TransactionConversionsSpec.scala:158]: maybe worth adding a test for consuming vs non-consuming?

**Review by cocreature — COMMENTED** [review]
Mostly looks good but we should clarify the semantics and maybe rename some stuff before merging. Thanks!

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala (hunk 4 above)
Comment (hubert-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala]: The idea was to follow the same idea as in Scala but you're right, maybe it's not clear. Changed.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala (hunk 1 above)
Comment (hubert-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala]: Agreed, thanks. Changed.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala (hunk 5 above)
Comment (hubert-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala]: Changed.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala (hunk 6 above)
Comment (hubert-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala]: I just inlined it, as it's trivial and used just once.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala [hunk 8]
```diff
@@ -92,9 +95,157 @@ object TransactionConversions {
...
+        }
+    }
 }
 
 final case class TransactionNodeIdWithNode(
     nodeId: RawTransaction.NodeId,
     node: RawTransaction.Node,
 )
+
+sealed trait ContractIdOrKey extends Product with Serializable
```
[the remark this answers is not in the document]
Comment (hubert-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala]: Done.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala [hunk 9]
```diff
@@ -201,8 +200,15 @@ object KeyValueCommitting {
 
       case DamlSubmission.PayloadCase.TRANSACTION_ENTRY =>
         val transactionEntry = submission.getTransactionEntry
-        transactionOutputs(transactionEntry) +
-          commandDedupKey(transactionEntry.getSubmitterInfo)
+        TransactionConversions
+          .decodeContractIdsAndKeys(RawTransaction(transactionEntry.getRawTransaction))
```
[the remark this answers is not in the document]
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala]: Good idea, done.

File: daml-lf/kv-support/src/test/scala/com/daml/lf/kv/transactions/TransactionConversionsSpec.scala (hunk 7 above)
Comment (hubert-da) [reply daml-lf/kv-support/src/test/scala/com/daml/lf/kv/transactions/TransactionConversionsSpec.scala]: Added, however not sure, as we don't treat them differently. Maybe will help someone in the future :)

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala (hunk 2 above)
Comment (hubert-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:152]: Looks like it's the case. But TBH, I don't know, and it would be great to get more explanation on that. Maybe @miklos-da knows the details?

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala [hunk 10]
```diff
@@ -92,9 +96,161 @@ object TransactionConversions {
       .sequence_
       .map(_ => RawTransaction(transactionBuilder.build.toByteString))
   }
+
+  /** Decodes and extracts outputs of a submitted transaction, that is the IDs and keys of contracts created or updated
+    * by processing a submission.
```
Comment (cocreature) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:101]: I think it’s worth documenting that we overapproximate the outputs meaning we include transient contracts, contract ids that only cause divulgence, keys that are not modified, everything under a rollback node, ….

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala [hunk 11]
```diff
@@ -92,9 +96,161 @@ object TransactionConversions {
...
+                    }
+                  }
+                case (Left(error), _) => Left(error)
+              }
+          }
+          .left
+          .map(ConversionError.DecodeError)
+    }
+
+  /** Removes `Fetch`, `LookupByKey` and `Rollback` nodes from a transaction tree. */
```
Comment (cocreature) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:176]:
```suggestion
  /** Removes `Fetch`, `LookupByKey` and `Rollback` nodes (including their children) from a transaction tree. */
```

**Review by cocreature — APPROVED** [review]
Thanks!

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala (hunk 10 above)
Comment (hubert-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:101]: Done, I mostly used a comment from below. @cocreature, are you able to help me answer a few questions?
1. Just to make sure, transient contracts are those created and archived within the same transaction, right?
2. Is it correct to say "contract IDs that may cause divulgence", what does it actually mean? Or is it more precise to say "contract IDs of contracts that may be divulged"?
3. Why do we keep "keys that are not modified". Is it just because it's easy or we cannot even say whether they were modified or not?
4. Does "under a rollback node" mean children of a rollback node? I assume we include them because of using all nodes from `transaction.getNodesList`, is that right?

Comment (cocreature) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:101]: 1. right
2. we divulge the contract payload not the contract id so arguably the first phrasing is correct.
3. I don’t know, that’s something for your team. But I suspect it’s the same reason why we overapproximate the other stuff: it doesn’t do much harm and changing it adds complexity.
4. Yes, children (or transitive children) of rollback nodes. And your analysis is exactly right.

Comment (meiersi-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:101]: @hubert-da : on "keys that are not modified" -- including these can lead to spurious contention between transactions; and we did observe that for one of our customers which uses keys to maintain some intra-transaction state. I have a vague memory that @miklos-da implemented a post-processing step, which removes writes to keys that are equal to their read.

In general, our code should aim to minimize accidental contention as that negatively impacts our customer's operating experience.

@hubert-da @cocreature : wrt "contract IDs that may cause divulgence" -- these MUST be included in the current state of the world, as the kv-state tracks which contracts have been divulged to whom; and divulgence thus results in write-conflicts. I believe we can change this now, as we use transaction local divulgence (i.e., redivulge all contracts in every transaction independent of the previous state of divulgence). However, it would be better for @miklos-da to coordinate this, as he knows the detailed state of the various changes.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala (hunk 3 above)
Comment (meiersi-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:147]: AFAIU https://docs.daml.com/concepts/ledger-model/ledger-privacy.html, non-consuming exercises and fetches behave very similary, and both can lead to divulgence. As explained above, divulgence can currently lead to writes to the kv-state.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala (hunk 10 above)
Comment (cocreature) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:101]: divulgence makes sense to me, including transient contracts on the other hand not really.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala [hunk 12]
```diff
@@ -92,9 +96,161 @@ object TransactionConversions {
...
   }
+
+  /** Decodes and extracts outputs of a submitted transaction, that is the IDs and keys of contracts created or updated
+    * by processing a submission.
+    *
+    * The results, among the others, may include contract IDs of transient contracts (created and archived within the same transaction),
+    * contract IDs that may cause divulgence and keys that are not modified. Actual outputs must be a subset of,
+    * or the same as, computed outputs and we currently relax this check by widening the latter set,
+    * treating a node the same regardless of whether it was a child of a rollback node or not, for example.
+    * Computed outputs that are not actual outputs can be safely trimmed.
```
Comment (meiersi-da) [inline daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:107]: Reading the comments by @cocreature (thanks for the detailed review!) and the commentary here, I believe it would be valuable to make an attempt to clarify further what exactly the "transaction output" are that are computed by this function. The comment as-is helps only partially as it references several concepts that are undefined a this dependency level of code; e.g. "actual outputs", "computed outputs", "safely trimmed", and likely also "processing a submission".

AFAIS, one viable definitions would be that that the outputs of a transaction `tx` are exactly (1) the contract-ids of contracts that `tx` archives or divulges and (2) the keys whose "key state" (as per https://docs.daml.com/concepts/ledger-model/ledger-integrity.html) is changed by `tx`.

There are other definitions possible, but I believe we should aim for a clean definition built on top of the ledger model here to avoid miscommunication between the language team and the kv team.

What do you think @hubert-da @miklos-da @cocreature @remyhaemmerle-da ? Can we find such a definition for the semantics required by the kv team?

Comment (cocreature) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:107]: Definitely in favor of cleaning this up and we were planning to do that as a follow-up. This PR was only reshuffling code so I just wanted to document the current semantics rather than changing them.

The semantics you suggested mostly sound fine to me but unless I’m missing something, I think we should exclude contracts created in the current transaction that are not active at the end. So slightly rephrased:

1. A contract id is an output if  one of the following is true:
      1. The transaction contains a fetch (potentially by key) or exercise for that contract id and the contract was not created in the current transaction. This captures divulgence and includes archives. Note that a lookup by key does not divulge (atm, I recall us wanting to change that but afaict we have not done so yet) so I’d not include the contract id for a successful lookup by key.
      2. The transaction contains a create for that contract id and the contract is still active at the end of the transaction. This excludes transient contracts and rolled back creates.
 2. Your key state definition sounds fine for keys.

Comment (meiersi-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:107]: Actually, after doing some further code spelunking, it looks to me that this is the only real use-site for the `extractTransactionOutputs` function: https://github.com/digital-asset/daml/blob/9ed1175d41237262fd195f7519ff512bf64e32c9/ledger/participant-state/kvutils/src/test/lib/scala/com/daml/ledger/participant/state/kvutils/KVTest.scala#L388

It is not clear to me why the check needs to be done against the result of `extractTransactionOutputs`. I'd suggest to change the usesite to check that
1. the read-set of the execution result is a subset of the inputs that were declared in the submission
2. the outputs are a subset of the read-set, which ensures that there is no write to a kv-key whose value has not been read. 

This way we can remove the need for an extra function and make the test more precise.

What do you think @hubert-da and @miklos-da ?

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala (hunk 2 above)
Comment (miklos-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:152]: @hubert-da I'd suggest figuring out who made the last changes to this section and asking that person — unfortunately, I don't have context either.

File: daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala (hunk 12 above)
Comment (hubert-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:107]: Thanks! I'll try to clean this up.

Comment (hubert-da) [reply daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala:107]: @meiersi-da, the `extractTransactionOutputs`/`transactionOutputs` function is also being used by `KeyValueCommitting.verifyStateUpdatesAgainstPreDeclaredOutputs` that on the other hand is being used by `KeyValueCommitting.processSubmission`.

@miklos-da, is there anything we'd like to do with that? Based on your comment and the above, I don't think we do.
